src/lib/mongoBackend/safeMongo.cpp

```diff
@@ -195,3 +198,117 @@ BSONElement getField(const BSONObj& b, const std::string& field)
   LM_E(("Runtime Error (field '%s' is missing in BSONObj <%s>)", field.c_str(), b.toString().c_str()));
   return BSONElement();
 }
+
+/* ****************************************************************************
+*
+* moreSafe -
+*
+* This is a safe version of the more() method for cursors in order to avoid
+* exception that may crash the broker. However, if the more() method is returning
+* an exception something really bad is happening, it is considered a Fatal Error.
+*
+* (The name resembles the same relationship between next() and nextSafe() in the
+* mongo driver)
+*
+*/
+bool moreSafe(const std::auto_ptr<DBClientCursor>& cursor)
+{
+  try
+  {
+    return cursor->more();
+  }
+  catch (std::exception &e)
+  {
+    LM_E(("Fatal Error (more() exception: %s)", e.what()));
+    return false;
+  }
+  catch (...)
+  {
+    LM_E(("Fatal Error (more() exception: generic)"));
+    return false;
+  }
+}
+
+/* ****************************************************************************
+*
+* nextSafeOrError -
+*
+*/
+bool nextSafeOrError(const std::auto_ptr<DBClientCursor>& cursor, BSONObj* r, std::string* err)
+{
+  try
+  {
+    *r = cursor->nextSafe();
+    return true;
+  }
+  catch (const std::exception &e)
+  {
+    *err = e.what();
+    return false;
+  }
+  catch (...)
+  {
+    *err = "generic exception";
+    return false;
+  }
+}
+
+/* ****************************************************************************
+*
+* safeGetSubId -
+*
+*/
+bool safeGetSubId(const SubscriptionId& subId, OID* id, StatusCode* sc)
+{
+  try
+  {
+    *id = OID(subId.get());
+    return true;
+  }
+  catch (const AssertionException &e)
+  {
+    // FIXME: this check is "defensive", but from a efficiency perspective this should be short-cut at
+    // parsing stage. Check it.
```

Comment:
Program flow ends in this catch() when the subscriptionId doesn't conform with OID format (i.e. 24 hex digits). What I mean is that this could be checked earlier, just after parsing the request, instead of doing the check in the deepness of mongoBackend.

Is the check done at parsing level? I don't know. And I don't mind :) at this moment, as long as we keep the FIXME to remember this in the future.